src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs

```diff
@@ -450,49 +456,50 @@ private async Task AddHttp11ConnectionAsync(HttpRequestMessage request)
                 }
                 catch (OperationCanceledException oce) when (oce.CancellationToken == cts.Token)
                 {
-                    HandleHttp11ConnectionFailure(request, CreateConnectTimeoutException(oce));
+                    HandleHttp11ConnectionFailure(queueItem.Waiter, CreateConnectTimeoutException(oce));
                     return;
                 }
                 catch (Exception e)
                 {
-                    HandleHttp11ConnectionFailure(request, e);
+                    HandleHttp11ConnectionFailure(queueItem.Waiter, e);
                     return;
                 }
             }
 
             // Add the established connection to the pool.
-            ReturnHttp11Connection(connection, isNewConnection: true);
+            ReturnHttp11Connection(connection, isNewConnection: true, queueItem.Waiter);
         }
 
         private void CheckForHttp11ConnectionInjection()
         {
             Debug.Assert(HasSyncObjLock);
 
-            if (!_http11RequestQueue.TryPeekUncanceledRequest(this, out HttpRequestMessage? request))
-            {
-                return;
-            }
+            _http11RequestQueue.PruneCompletedRequestsFromHeadOfQueue(this);
 
             if (NetEventSource.Log.IsEnabled())
             {
                 Trace($"Available HTTP/1.1 connections: {_availableHttp11Connections.Count}, Requests in the queue: {_http11RequestQueue.Count}, " +
+                    $"Requests without a connection attempt: {_http11RequestQueue.RequestsWithoutAConnectionAttempt}, " +
                     $"Pending HTTP/1.1 connections: {_pendingHttp11ConnectionCount}, Total associated HTTP/1.1 connections: {_associatedHttp11ConnectionCount}, " +
                     $"Max HTTP/1.1 connection limit: {_maxHttp11Connections}.");
             }
 
             // Determine if we can and should add a new connection to the pool.
             if (_availableHttp11Connections.Count == 0 &&                           // No available connections
                 _http11RequestQueue.Count > _pendingHttp11ConnectionCount &&        // More requests queued than pending connections
-                _associatedHttp11ConnectionCount < _maxHttp11Connections)           // Under the connection limit
+                _associatedHttp11ConnectionCount < _maxHttp11Connections &&         // Under the connection limit
+                _http11RequestQueue.RequestsWithoutAConnectionAttempt != 0)         // There are requests we haven't issued a connection attempt for
             {
                 _associatedHttp11ConnectionCount++;
                 _pendingHttp11ConnectionCount++;
 
-                Task.Run(() => AddHttp11ConnectionAsync(request));
+                RequestQueueItem<HttpConnection> queueItem = _http11RequestQueue.PeekNextRequestForConnectionAttempt();
+
+                Task.Run(() => AddHttp11ConnectionAsync(queueItem));
```

Comment:
This change is also avoiding a closure allocation that seems to have snuck into CheckForHttp11ConnectionInjection.